### DISCUSSION TITLE
Waveshare RS485 Connection Method

### DISCUSSION THREAD

**Discussion by Daz42** [nathanmarlor/foxess_modbus#329] [Support] [answered]
Hi there, firstly thank you for this integration. I was using it over the LAN connection to my H1. As I want to control the BMS, I purchased the Waveshare POE server and followed the directions precisely to set it up. I have wired 485A to pin 4 and 485B to pin3 in the inverter. Also used a multi meter to ensure I had a good connection. When setting up the modes in HA, using the correct IP address and the port, I am getting:

Error connecting to your adapter or inverter. Ensure the adapter is properly configured and is correctly wired to your inverter (see the setup link above), then try again. Details: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received).

I think I'm getting a TCP connection to the Waveshare but may not be getting any data from the inverter. Any ideas?

Thank you!

Daz.

**Comment by canton7** [comment]
> I think I'm getting a TCP connection to the Waveshare but may not be getting any data from the inverter

That's what it looks like to me as well. What you've described looks correct also.

Check the slave ID on the inverter is correct (Settings -> Communication -> RS485 -> Device ID) -- it should be 247. Other than that I'm not sure I've got anything to suggest apart from the obvious: make sure that all of the settings are correct, that you've definitely got the IP of the waveshare, that you didn't break a connection or wire when you plugged the connector back into the inverter, etc. Most of these sorts of questions seem to be resolved with "A wire fell out" or similar.

**Comment by FozzieUK** [comment]
If you were using the LAN connection, you need to disable the ethernet on your inverter before the RS485 will work.

Common mistakes are the RS485A(4) must be wired to A of the device, and RS485B(3) to B and also with the Waveshare there are often settings errors, always double check them, particularly those which relate to TCP_Server, Transfer protocol being Modbus_TCP to RTP, and the device port being 502.

It is best to test using modpoll first from a laptop or desktop connected the the same subnet,

You can download that from here if you don’t have it -
https://www.modbusdriver.com/modpoll.html

Then issue this command at the command prompt -
modpoll.exe -m tcp -p 502 -a 247 -t 3 -0 -1 -c 16 -r 11000 192.168.x.x (<<< replace this with your waveshare IP address)

And hopefully you should get a response

**Reply by Daz42** [reply]
Thank you for that! I looked at my H1 menus and there doesn't seem to be an option to disable the LAN port - it was always enabled and all I did was configure a static IP address. There is a remote access option but not sure if it's related.

**Reply by FozzieUK** [reply]
There used to be a disable/enable option but it’s been quite a while since i’ve been into the communications, Ethernet menu, i’ll double check that later when i’m at my inverter.

As Antony said, check your slave is set at 247, and all the waveshare settings and then see what result you get from the modpoll 👍

**Reply by Daz42** [reply]
Under communications, there are no settings to disable the LAN port - at least on my H1. I've downloaded ModPoll and run it using the command line you suggested (with my IP address set). It looks like I am reaching the Waveshare but getting this error message:

PS D:\Software\Home Assistant\FoxESS Integration\Modpoll3.10\win> .\modpoll.exe -m tcp -p 502 -a 247 -t 3 -0 -1 -c 16 -r 11000 192.168.1.150
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 16
Communication.........: 192.168.1.150, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
Reply time-out!

So this looks like the Inverter isn't responding. I've checked the wiring to pin 4 (485A) and pin 3(485B) with a multimeter using the grub screws as contact points to ensure that there is a solid wire grab and all looks good there.

**Comment by Daz42** [comment]
Ok, I found the issue. I have two Waveshares - the other is for my other property subject to this solution working. I swapped them out and it's working. Thank you very much both Antony and Dave. I am now able to control my BMS settings - I really didn't want to plug in the WiFi adapter again! So the bit we learned in addition to the solution is that the H1 6.0 E version 0.59 does not have a LAN port enable/disable and does affect the Waveshare integration using 485A and 485B.

**Reply by canton7** [reply]
Good news! Sounds like it was some configuration on the Waveshare then?

> the H1 6.0 E version 0.59 does not have a LAN port enable/disable and does affect the Waveshare integration using 485A and 485B.

What exactly do you mean by that?

**Comment by Daz42** [comment]
It was a fault Waveshare - all of the settings were correct and I had reset to default to check before configuring again. I'll return the bad one. Reference to the LAN port to discussion with Dave as he had suggested that the LAN port had to be disabled before using the 485 pins - but in my case there wasn't a setting and the 485 pins worked regardless.

**Reply by FozzieUK** [reply, marked as answer]
Great, glad you’ve got it working ok 👍 

Before you send the faulty one back, if you have updated the firmware it has been known to scramble it’s config - if you put a link across then NC pins for more than 5 seconds, it will factory default and ‘usually’ it works afterwards :smiley:

**Reply by Daz42** [reply]
Good info, I will give that a try. Not surprising with value Chinese kit - a reset might work!

**Reply by Daz42** [reply]
Ok, that worked Dave. Both Waveshares work now. I'm super grateful for your help.

**Reply by canton7** [reply]
Nice one!

I've added a Troubleshooting section to the wiki page with this info.

**Reply by FozzieUK** [reply]
Brilliant 👍 , i’ve been really impressed with them for the money.
@canton7 thanks Antony, I was going to do that when I got back later, there’s been a few that have suffered from this ☺️

**Reply by ShornOfIdeas** [reply]
Hi @FozzieUK. Having a similar issue here so extending this thread rather than opening a new one, hoping that is correct etiquette!

My issue is the waveshare poe is not responding. The power light is on and when I insert the ethernet cable the Link light comes on too, but not the Act light. A is connected to pin 4 and B to 3 on the KH modbus connector. I've triple checked the settings in Vircom and I've tried updating the firmware but there is no connection so naturally I get an error on that. Vircom confirms the IP address, port (502) and KH slave (247) as expected but shows the waveshare as unresponsive. 

Modpoll says:
_modpoll 3.16 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2025 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 16
Communication.........: 192.168.1.200, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

Can't reach server/slave! Check TCP/IP and firewall settings._

I've switched firewall off, but no change. I've run an ethernet cable direct from router, but that's not it either (I'm otherwise using TPLink ethernet over power).  TCP/IP settings as follows:

Windows IP Configuration

   Host Name . . . . . . . . . . . . : DESKTOP-J151IK5
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No

Unknown adapter Secure VPN:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : AVG Secure VPN Wintun Adapter
   Physical Address. . . . . . . . . :
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

Wireless LAN adapter Local Area Connection* 1:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft Wi-Fi Direct Virtual Adapter #3
   Physical Address. . . . . . . . . : 9E-30-5B-A0-38-0D
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Wireless LAN adapter Local Area Connection* 2:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft Wi-Fi Direct Virtual Adapter #4
   Physical Address. . . . . . . . . : AE-30-5B-A0-38-0D
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

Wireless LAN adapter Wi-Fi:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Qualcomm QCA61x4A 802.11ac Wireless Adapter
   Physical Address. . . . . . . . . : 9C-30-5B-A0-38-0D
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::43b4:507e:e7c0:21c8%19(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.0.21(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : 07 July 2025 13:20:38
   Lease Expires . . . . . . . . . . : 09 July 2025 11:12:45
   Default Gateway . . . . . . . . . : 192.168.0.1
   DHCP Server . . . . . . . . . . . : 192.168.0.1
   DHCPv6 IAID . . . . . . . . . . . : 362557531
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-23-F4-CC-3A-9C-30-5B-A0-38-0D
   DNS Servers . . . . . . . . . . . : 194.168.4.100
                                       194.168.8.100
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Bluetooth Network Connection:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Bluetooth Device (Personal Area Network)
   Physical Address. . . . . . . . . : 9C-30-5B-A0-38-0E
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Any thoughts gratefully received. Many thanks

**Reply by ShornOfIdeas** [reply]
I asked Gemini and it identified a subnet mask mismatch. Now all sorted :)

**Reply by FozzieUK** [reply]
Ahh yep I can see that, you had mixed two subnets .1.x and .0.x - bravo Gemini 😄

**Comment by ShornOfIdeas** [comment]
Any thanks for checking it over Dave. So, all I need to do now is work out
how to set up an automation. Just a simple force charge if SOC falls below
30%.  I don't find it's intuitively obvious! I've been searching for an
idiot's guide but I haven't found it yet. Cheers, Sean

**Reply by FozzieUK** [reply]
They are not the easiest to learn, the best way to do what you want is to create a new automation from scratch, in the 'when' box add a trigger 'Entity' pick 'Numeric State' then in the dialogue that pops up enter the name of the entity you want i.e. 'battery soc' and at then in the 'Below' box enter 30.
At the bottom in the 'Then Do', click 'add action' in the search box type 'select' and pick the 'Select: Select' in the list, it will give you a dialogue, click the 'Entity' button and start typing 'workmode'and click on the entity, then in the 'Option' box type 'Force Charge' (you must enter that precisely as written).

Give it a name and save it, then whenever your battery soc goes below 30% it will change the workmode to force charge.

Don't forget to change it back when it gets to your desired maximum, probably easier to do that in another automation but instead of 'Force Charge' use 'Self Use' or 'Feed-in First'

You can create a helper toggle and use it in the 'And If' section of the automation and use that to override the behaviour i.e. create a toggle helper called 'enable force charge' then in the and if section add an action of Entity, choose 'State' in the entity name box type the 'enable force charge' in the 'state' box type 'On' and then it will only activate when the enable force charge is on - and add that entity to your dashboard somewhere so you can toggle the switch.

**Comment by ShornOfIdeas** [comment]
Thanks Dave - I would probably never have worked that out - so thank you
again! Is there a "how to guide" anywhere?

I added in an "and if" to restrict charging to the low tariff period of
04:00 - 07:00. At some point I'll try and add in weather dependencies, e.g.
if daytime forecast is cloudy and SOC is below 25 then charge between 13:00
and 16:00.

Cheers, Sean

**Comment by ShornOfIdeas** [comment]
Hi Dave
After quite a gap, the automation (charge if SOC less than 30%) triggered
this morning. The problem is the off automation (above 50%) is not working.
The system is on permanent charge. Getting expensive!

This is the set up for above 50% (I've also tried irt with "Self Use" but
that does not activate either. Any ideas please? Many thanks, SEan

[image: image.png]

**Reply by FozzieUK** [reply]
Hi, sorry I can’t see any attachment, can you copy a screen shot of the automation below

**Comment by ShornOfIdeas** [comment]
Hi Dave and thank you for responding. Here is the screenshot of what I have
(jpg also attached). I set this up yesterday morning but it failed to
trigger. So, I then switched manually to "Self Use" from the dashboard.

Strangely, the dashboard shows that new automation attached here then
triggered overnight, at about 1am, when the SOC would have been about 85%
and decreasing from a higher level. So completely random compared with what
I am attempting to achieve (trigger "Self Use" if charging increases SOC
above 50%)

Any thoughts on what I'm doing wrong here? Many thanks for your help.
Regards, Sean.

[image: image.png]

**Reply by FozzieUK** [reply]
I still can't see the image it just says [image: image.png], I think it must be the way you are adding it - you'll need to go to github and paste it or drop it on the page.

Alternatively, got to the automation and open it in the editor and in the corner you'll see the 3 dots, click that and select 'Edit in YAML' that will show you what the automation text is - select it all and copy and paste the text below.

**Comment by ShornOfIdeas** [comment]
Sorry about this. Here is the code:

alias: Turn off >50
description: ""
triggers:
  - trigger: numeric_state
    entity_id:
      - sensor.foxess_bat_soc
    above: 50
conditions: []
actions:
  - action: select.select_option
    metadata: {}
    data:
      option: Self Use
    target:
      entity_id: select.work_mode
mode: single

Thanks, Sean

**Reply by kharas80** [reply]
Maybe it is also helpful to post your trace Information about this automation.

**Reply by FozzieUK** [reply]
@ShornOfIdeas 
I have a suspicion this is because of the way a numeric test on the battery soc entity is behaving as it is a string i.e. the entity does not contain 50 but "50" instead

Open your automation and in the battery soc test at the bottom you will see a 'Value template' field, enter this text into template

 `{{ state.state|int }}`

i.e. your test should look like 

![111](https://github.com/user-attachments/assets/98cd4423-af32-4029-b366-534517b9aa16)

What this does is makes sure the state of the state object is converted to an integer before the test takes place - tested on my system and works fine.

Add this text in your 30% and your 50% automations and hopefully that should working correctly.

**Comment by ShornOfIdeas** [comment]
Thanks again Dave. I made that change and reset the parameters such that
the "force charge" would be triggered immediately - but nothing happened.
Mysterious! Thanks for your help with this. Regards, Sean

**Reply by FozzieUK** [reply]
The test should work ok (give it time), but the way a trigger test works may be confusing you - a trigger test has to go from false to true by the entity changing for it to fire. 
So for example it wouldn't fire if it were already over 50 it must have been below the test limit and go above it for the test to fire, the upside being that it doesn't fire all the time when it is above the test, just the once when it becomes true.

You can make it work so that it will always fire straight away, you'll need to change your trigger to a time pattern trigger with a pattern of '/1' in the minutes box (i.e. fire each and every minute)

Then move your battery soc > 50 test down to the 'And If' conditions, it might also be a good idea to add an entity test for work mode being 'Force Charge' to avoid it triggering all the time.

Doing it that way means the automation will trigger every minute, check it's conditions and if both are true proceed to the action.

**Comment by ShornOfIdeas** [comment]
Got it - thanks for explaining what should perhaps have been obvious. I
think I'll leave it as it is, as that is what I want to happen in the real
world. While we are in summer (with lots of sunlight for the PV) the
trigger conditions are infrequent, hence getting a bit jumpy over whether
things are working. Many thanks for all your help Dave :)

Best regards, Sean
